clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/provider/agent/CloudFoundrySpaceCachingAgent.java

```diff
@@ -0,0 +1,102 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.netflix.spinnaker.clouddriver.cloudfoundry.provider.agent;
+
+import static com.netflix.spinnaker.cats.agent.AgentDataType.Authority.AUTHORITATIVE;
+import static com.netflix.spinnaker.clouddriver.cloudfoundry.cache.Keys.Namespace.SPACES;
+import static java.util.Collections.emptyMap;
+import static java.util.stream.Collectors.toSet;
+
+import com.fasterxml.jackson.databind.MapperFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.netflix.spectator.api.Registry;
+import com.netflix.spinnaker.cats.agent.AgentDataType;
+import com.netflix.spinnaker.cats.agent.CacheResult;
+import com.netflix.spinnaker.cats.agent.DefaultCacheResult;
+import com.netflix.spinnaker.cats.cache.CacheData;
+import com.netflix.spinnaker.cats.provider.ProviderCache;
+import com.netflix.spinnaker.clouddriver.cloudfoundry.cache.Keys;
+import com.netflix.spinnaker.clouddriver.cloudfoundry.cache.ResourceCacheData;
+import com.netflix.spinnaker.clouddriver.cloudfoundry.model.CloudFoundrySpace;
+import com.netflix.spinnaker.clouddriver.cloudfoundry.security.CloudFoundryCredentials;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+@Getter
+@Slf4j
+public class CloudFoundrySpaceCachingAgent extends AbstractCloudFoundryCachingAgent {
+
+  private static final ObjectMapper cacheViewMapper =
+      new ObjectMapper().disable(MapperFeature.DEFAULT_VIEW_INCLUSION);
+
+  private final Collection<AgentDataType> providedDataTypes =
+      Arrays.asList(AUTHORITATIVE.forType(SPACES.getNs()));
+
+  public CloudFoundrySpaceCachingAgent(CloudFoundryCredentials credentials, Registry registry) {
+    super(credentials, registry);
+  }
+
+  @Override
+  public CacheResult loadData(ProviderCache providerCache) {
+    long loadDataStart = this.getInternalClock().millis();
+    String accountName = getAccountName();
+    log.info("Caching all spaces in Cloud Foundry account " + accountName);
+
+    List<CloudFoundrySpace> spaces = getCredentials().getSpacesLive();
+
+    Map<String, Collection<CacheData>> results =
+        io.vavr.collection.HashMap.<String, Collection<CacheData>>of(
+                SPACES.getNs(),
+                spaces.stream()
+                    .map(
+                        s ->
+                            new ResourceCacheData(
+                                Keys.getSpaceKey(accountName, s.getRegion()),
+                                cacheView(s),
+                                emptyMap()))
+                    .collect(toSet()))
+            .toJavaMap();
```

Comment:
Try Guava's ImmutableMap.of()... It's much more standard than this vavr thing, plus has the benefit of immutability.

(Looking around I see this vavr thing is used all over CloudFoundry for some reason, but Guava is much, much more common in the rest of Spinnaker/the world, so we should prefer that.)

Reply:
@plumpy done!